src/material-date-fns-adapter/adapter/date-fns-adapter.ts

```diff
@@ -0,0 +1,238 @@
+      if (dtf) {
+        // date-fns doesn't appear to support this functionality.
+        // Fall back to `Intl` on supported browsers.
+        const date = new Date();
+        date.setUTCFullYear(2017, 0, i + 1);
+        date.setUTCHours(0, 0, 0, 0);
+        return dtf.format(date).replace(/[\u200e\u200f]/g, '');
+      }
+
+      return i + '';
+    });
+  }
+
+  getDayOfWeekNames(style: 'long' | 'short' | 'narrow'): string[] {
+    const pattern = DAY_OF_WEEK_FORMATS[style];
+    return range(7, i => this.format(new Date(2017, 0, i + 1), pattern));
+  }
+
+  getYearName(date: Date): string {
+    return this.format(date, 'y');
+  }
+
+  getFirstDayOfWeek(): number {
+    return this.locale.options?.weekStartsOn ?? 0;
+  }
+
+  getNumDaysInMonth(date: Date): number {
+    return getDaysInMonth(date);
+  }
+
+  clone(date: Date): Date {
+    return new Date(date.getTime());
+  }
+
+  createDate(year: number, month: number, date: number): Date {
+    if (typeof ngDevMode === 'undefined' || ngDevMode) {
+      // Check for invalid month and date (except upper bound on date which we have to check after
+      // creating the Date).
+      if (month < 0 || month > 11) {
+        throw Error(`Invalid month index "${month}". Month index has to be between 0 and 11.`);
+      }
+
+      if (date < 1) {
+        throw Error(`Invalid date "${date}". Date has to be greater than 0.`);
+      }
+    }
+
+    // Passing the year to the constructor causes year numbers <100 to be converted to 19xx.
+    // To work around this we use `setFullYear` and `setHours` instead.
+    const result = new Date();
+    result.setFullYear(year, month, date);
+    result.setHours(0, 0, 0, 0);
+
+    // Check that the date wasn't above the upper bound for the month, causing the month to overflow
+    if (result.getMonth() != month && (typeof ngDevMode === 'undefined' || ngDevMode)) {
+      throw Error(`Invalid date "${date}" for month with index "${month}".`);
+    }
+
+    return result;
+  }
+
+  today(): Date {
+    return new Date();
+  }
+
+  parse(value: any, parseFormat: string | string[]): Date | null {
+    if (typeof value == 'string' && value.length > 0) {
+      const iso8601Date = parseISO(value);
+
+      if (this.isValid(iso8601Date)) {
+        return iso8601Date;
+      }
+
+      const formats = Array.isArray(parseFormat) ? parseFormat : [parseFormat];
+
+      if (!parseFormat.length) {
+        throw Error('Formats array must not be empty.');
+      }
+
+      for (const currentFormat of formats) {
+        const fromFormat = parse(value, currentFormat, new Date());
+
+        if (this.isValid(fromFormat)) {
+          return fromFormat;
+        }
+      }
+
+      return this.invalid();
+    } else if (typeof value === 'number') {
+      return new Date(value);
+    } else if (value instanceof Date) {
+      return this.clone(value);
+    } else if (value != null && value !== '') {
+      return new Date(Date.parse(value));
```

Comment:
What's the case in which we'd want to fall back to the native `Date` parsing? Would it be better to throw an error?

Reply:
I added this to be a bit more defensive and to handle any values that weren't picked up by the previous blocks. The majority of cases should be picked up further up.

Reply:
My concern is that, from the developer's perspective, it's probably more predictable to treat this as an invalid date rather than fall back to the nonstandard, US-centric behavior of `Date.parse`. If I was explicitly deciding to use `date-fns` for my parsing logic, I would be surprised if there was a case falling back to the native parsing.

Reply:
Alright, I'll remove it for now. As I mentioned above, it's unlikely that we would've even hit that case.